node/basicnode/map_test.go

```diff
@@ -46,3 +51,277 @@ func BenchmarkSpec_Unmarshal_Map3StrInt(b *testing.B) {
+	errExpect = `func called on wrong kind: "AssignString" called on a map node \(kind: map\), but only makes sense on string`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	err = b.AssignNode(basicnode.NewInt(3))
+	errExpect = `func called on wrong kind: "AssignNode" called on a map node \(kind: int\), but only makes sense on map`
+	qt.Check(t, err, qt.ErrorMatches, errExpect)
+
+	// TODO(dustmop): BeginList, AssignNull, AssignBytes, AssignLink
+}
+
+// Test that the map builder can create map nodes, and AssignNode will copy
+// such a node, and lookup methods on that node will work correctly
+func TestMapBuilder(t *testing.T) {
+	b := basicnode.Prototype.Map.NewBuilder()
+
+	// construct a map of three keys, using the MapBuilder
+	ma, err := b.BeginMap(3)
+	if err != nil {
+		t.Fatal(err)
+	}
+	a := ma.AssembleKey()
+	a.AssignString("cat")
+	a = ma.AssembleValue()
+	a.AssignString("meow")
+
+	a, err = ma.AssembleEntry("dog")
+	if err != nil {
+		t.Fatal(err)
+	}
+	a.AssignString("bark")
+
+	a = ma.AssembleKey()
+	a.AssignString("eel")
+	a = ma.AssembleValue()
+	a.AssignString("zap")
+
+	err = ma.Finish()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// test the builder's prototypes and its key and value prototypes, while we're here
+	np := b.Prototype()
+	qt.Check(t, fmt.Sprintf("%T", np), qt.Equals, "basicnode.Prototype__Map")
+	np = ma.KeyPrototype()
+	qt.Check(t, fmt.Sprintf("%T", np), qt.Equals, "basicnode.Prototype__String")
+	np = ma.ValuePrototype("")
+	qt.Check(t, fmt.Sprintf("%T", np), qt.Equals, "basicnode.Prototype__Any")
+
+	// compare the printed map
+	mapNode := b.Build()
+	actual := printer.Sprint(mapNode)
+
+	expect := `map{
+	string{"cat"}: string{"meow"}
+	string{"dog"}: string{"bark"}
+	string{"eel"}: string{"zap"}
+}`
+	qt.Check(t, expect, qt.Equals, actual)
+
+	// copy the map using AssignNode
+	c := basicnode.Prototype.Map.NewBuilder()
+	err = c.AssignNode(mapNode)
+	if err != nil {
+		t.Fatal(err)
+	}
+	anotherNode := c.Build()
+
+	actual = printer.Sprint(anotherNode)
+	qt.Assert(t, expect, qt.Equals, actual)
+
+	// access values of map, using string
+	r, err := anotherNode.LookupByString("cat")
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Check(t, "meow", qt.Equals, must.String(r))
+
+	// access values of map, using node
+	r, err = anotherNode.LookupByNode(basicnode.NewString("dog"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Check(t, "bark", qt.Equals, must.String(r))
+
+	// access values of map, using PathSegment
+	r, err = anotherNode.LookupBySegment(datamodel.ParsePathSegment("eel"))
+	if err != nil {
+		t.Fatal(err)
+	}
+	qt.Check(t, "zap", qt.Equals, must.String(r))
+
+	// validate the node's prototype
+	np = anotherNode.Prototype()
+	qt.Check(t, fmt.Sprintf("%T", np), qt.Equals, "basicnode.Prototype__Map")
+}
+
+// test that AssignNode will fail if called twice, it expects an empty
+// node to assign to
+func TestMapCantAssignNodeTwice(t *testing.T) {
+	b := basicnode.Prototype.Map.NewBuilder()
```

Comment:
So for almost all of these (I pick just one instance at random to comment on), I wonder if we can immediately generalize them to `TestTheTrait(t *testing.T, np datamodel.NodePrototype)` ?

It would need:
- the reusable test would need comments on it on what kind of NodePrototype it's expecting (in this case: a map)
- another quick func of boilerplate for `TestTheTrait(t *testing.T) { sharedtests.TestTheTrait(t,  basicnode.Prototype.Map) }`
- ... that's about it?

Not sure if you saw it, but we have the `go-ipld-prime//node/tests` package for some of these already.

Reply:
(It's also very likely that this will run into issues being applied on more than one node implementation, especially in the area of error message string matching... but I'd like to be aspirational about that!)